gnuradio-runtime/lib/block.cc

```diff
@@ -594,15 +596,54 @@ void block::reset_perf_counters()
 
 void block::system_handler(pmt::pmt_t msg)
 {
-    // std::cout << "system_handler " << msg << "\n";
-    pmt::pmt_t op = pmt::car(msg);
-    if (pmt::eqv(op, d_pmt_done)) {
-        d_finished = pmt::to_long(pmt::cdr(msg));
-        global_block_registry.notify_blk(d_symbol_name);
+    // check whether message is cons or dict
+    if (pmt::is_pair(msg)) {
+        // cons
+        // TODO: DEPRECATE sending pmt::cons in favor of dict
+        auto op = pmt::car(msg);
+
+        if (pmt::eqv(op, d_pmt_done)) {
+            auto value = pmt::cdr(msg);
+            // check whether the finishedness is bool or numeric
+            if (pmt::is_bool(value)) {
+                d_finished = pmt::to_bool(value);
+            } else {
+                try {
+                    // TODO DEPRECATE passing of numbers as finishedness
+                    d_finished = pmt::to_long(pmt::cdr(msg));
+                } catch (pmt::wrong_type& e) {
+                    std::cout << "WARNING: bad message value on system port: " << e.what()
+                              << "\n";
+                    return;
+                }
+            }
+        } else {
+            std::cout << "WARNING: bad message op on system port: " << op << "\n";
+        }
+    } else if (pmt::is_dict(msg)) {
+        // dict
+        auto value = pmt::dict_ref(msg, d_pmt_done, pmt::PMT_NIL);
+
+        if (value != pmt::PMT_NIL) {
+            try {
+                d_finished = pmt::to_bool(value);
+            } catch (pmt::wrong_type& e) {
+                std::cout << "WARNING: got a " << d_pmt_done
+                          << " key with non-bool value: " << e.what() << "\n";
+                return;
+            }
+        } else {
+            std::cout << "WARNING: bad message op in dict on system port!\n";
```

Comment:
is `d_pmt_done` the only thing this port can handle?

Reply:
yep, at this point that's true – but I'm sure the intention was to control more through that port. Honestly, don't know whether that's wise.